### DISCUSSION TITLE
Refresh and reload entities through session

### DISCUSSION THREAD

**Discussion by Cartman34** [doctrine/doctrine2#8766] [Support Questions]
Hello,

I am getting some trouble to maintain some code source with Doctrine entities through session.

While registering an user, he may register his enterprise with an establishment with an address, Also he may be having a personal address and in most cases, this address is the same than the establishment.
An address is unique in database, so we are re-using an existing entity if we find it in database.
All these entities may be persisted in session while using the wizard. Some of them can or can not be persisted in database, I can not predict it for all of them.

I keep db persisted entities in session to keep relations but in fact, it should be reloaded before each usage. To do that, I am using my own way with a simple find($id) and get the db object.
From manual, I thought EntityManager::refresh() may do the job but in fact no, it returns an error: "Undefined index: 000000004156b3a300000000288c6c93" while flushing (it seems the entity should be currently persisted)
When getting a flush error, all my session's entities are corrupted because they got an ID in PHP object but they are not persisted in database (Whaaat ?).

All this system is complex and I am not sure I am using Symfony & Doctrine in the right way.
What are the best practices in case of entities in session ? What should I do ?

**Comment by beberlei** [comment]
In my opinion you should not store the entity in the session, but convert it into either just the ID or all the other relevant information using a data transfer object.

**Reply by Cartman34** [reply]
Ok, like a reference object ? https://github.com/doctrine/orm/blob/2.2/lib/Doctrine/ORM/EntityManager.php#L374-413
It seems tedious to navigate through all entity hierarchy

**Reply by Cartman34** [reply]
https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/cookbook/entities-in-session.html
Their advice is to use detach() and __sleep()
I need an automatic way to replace related entity in hierarchy that are persisted in database.

**Reply by SenseException** [reply]
This part of the documentation needs an overhaul. Using DTOs or other value objects should be preferred to an entity.
